cli/flox-rust-sdk/src/models/environment/core_environment.rs

```diff
@@ -515,6 +570,66 @@ impl CoreEnvironment<ReadOnly> {
         })
     }
 
+    fn upgrade_with_catalog_client(
+        &mut self,
+        flox: &Flox,
+        client: &impl ClientTrait,
+        groups_or_iids: &[String],
+        manifest: &TypedManifestCatalog,
+    ) -> Result<UpgradeResult, CoreEnvironmentError> {
+        let mut existing_lockfile = 'lockfile: {
+            let Ok(lockfile_path) = CanonicalPath::new(self.lockfile_path()) else {
+                break 'lockfile None;
+            };
+            let lockfile = LockedManifest::read_from_file(&lockfile_path)
+                .map_err(CoreEnvironmentError::LockedManifest)?;
+            match lockfile {
+                LockedManifest::Catalog(lockfile) => Some(lockfile),
+                _ => {
+                    warn!(
+                        "Found version 1 manifest, but lockfile doesn't match: Ignoring lockfile."
+                    );
+                    None
+                },
+            }
+        };
+
+        if let Some(ref mut lockfile) = existing_lockfile {
+            let locked_packages = std::mem::take(&mut lockfile.packages);
+
+            lockfile.packages = locked_packages
+                .into_iter()
+                .filter(|pkg| {
+                    !(groups_or_iids.contains(&pkg.group)
+                        || groups_or_iids.contains(&pkg.install_id))
+                })
+                .collect();
+        }
+
+        let upgraded =
+            LockedManifestCatalog::lock_manifest(manifest, existing_lockfile.as_ref(), client)
```

Comment:
note nonblocking: a part of me wonders if it would make more sense to push the upgrade logic deeper into e.g. `collect_package_groups`. Maybe something to keep in mind when we do https://github.com/flox/flox/issues/1398

Reply:
I'd say the abstraction level for locking is right now, i.e. "lock with these packages already locked". Upgrade is only controlling the "packages already locked" part which i think makes sense.
Iirc think `lock_manifest` previously took a more processed form of the existing lockfile, but we changed that to process it within the method. I'd consider pulling _that_ out again and giving it a nicer interface to selectively unlock things (and maybe create resolution groups)